sql/handler.cc

```diff
@@ -3150,8 +3150,9 @@ int handler::update_auto_increment()
     */
     if (thd->killed == KILL_BAD_DATA ||
         nr > table->next_number_field->get_max_int_value())
-      DBUG_RETURN(HA_ERR_AUTOINC_ERANGE);
-
+      result= HA_ERR_AUTOINC_ERANGE;
```

Comment:
This doesn't make any sense to me. I think the bug is in `ha_partition::release_auto_increment()`. It makes certain incorrect assumptions and doesn't unlock autoinc lock when `handler::update_auto_increment()` fails.

And, please, add a descriptive commit comment next time so that I wouldn't need to debug your test case to understand what's going on.